packages/core/src/connectivity/scp-cf/xsuaa-service-xssec.ts

```diff
@@ -0,0 +1,85 @@
+import * as xssec from '@sap/xssec';
+import CircuitBreaker from 'opossum';
+import { parseSubdomain } from './subdomain-replacer';
+import { decodeJwt } from './jwt';
+import { ServiceCredentials } from './environment-accessor-types';
+import {
+  circuitBreakerDefaultOptions,
+  ResilienceOptions
+} from './resilience-options';
+
+type XsuaaServiceXssecCircuitBreaker = CircuitBreaker<
+  [requestUserTokenParam: RequestUserTokenParam],
+  string
+>;
+let circuitBreaker: XsuaaServiceXssecCircuitBreaker;
+function getCircuitBreaker(): XsuaaServiceXssecCircuitBreaker {
+  if (!circuitBreaker) {
+    circuitBreaker = new CircuitBreaker(
+      xssec.requests.requestUserToken,
+      circuitBreakerDefaultOptions
+    );
+  }
+  return circuitBreaker;
+}
+
+// `@sap/xssec` uses/checks `null` in general without considering `undefined`.
+interface SubdomainAndZoneId {
+  subdomain: string | null;
+  zoneId: string | null;
+}
+
+/**
+ * Get subdomain and zoneId value from a given JWT.
+ * @param jwt - A JWT from the current user.
+ * @returns subdomain and zoneId from the JWT
+ * @hidden
+ */
+export function getSubdomainAndZoneId(jwt?: string): SubdomainAndZoneId {
+  let subdomain: string | null = null;
+  let zoneId: string | null = null;
+  if (!jwt) {
+    return { subdomain, zoneId };
+  }
+
+  const decodedJwt = decodeJwt(jwt);
+  if (decodedJwt.iss) {
+    subdomain = parseSubdomain(decodedJwt.iss);
+  }
+  zoneId = decodedJwt.zid || zoneId;
+  return { subdomain, zoneId };
+}
```

Comment:
I checked the xssec library and it does not seem to be correct that it can only handle null. Let's discuss this, maybe I missed something.
Not using null could make this code a little more concise.

Also: we could consider changing `parseSubdomain` to allow and return `undefined` (or `null`) values to avoid the `if`s.

```suggestion
export function getSubdomainAndZoneId(jwt?: string): SubdomainAndZoneId {
  let subdomain: string | null = null;
  let zoneId: string | null = null;

  if (jwt) {
    const decodedJwt = decodeJwt(jwt);
    if (decodedJwt.iss) {
      subdomain = parseSubdomain(decodedJwt.iss);
    }
    if (decodedJwt.zid) {
      zoneId = decodedJwt.zid;
     }
  }
  return { subdomain, zoneId };
}
```

Reply:
Double checked. For the time being, 
- the `subdomain` works fine with both `null` and `undefined`.
- the `zoneId` works fine with `null` and `undefined`, but it will be set to `null` (`zoneId = null`) under some conditions, which might be risky in the future.
- the `scopes` works with `null`, because the source code checks `if(scopes !== null)`.
- some other parameters have `null` check like `if(abc !== null)`.

Therefore, I still think this brings some values.

Reply:
Agreed.